adapters/sspBC/sspbc.go

```diff
@@ -0,0 +1,401 @@
+		return adCreationError
+	}
+	// append bid to responses
+	bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+		Bid:     &bid,
+		BidType: bidType,
+	})
+
+	return nil
+}
+
+func getOriginalImpID(impID string, imps []openrtb2.Imp) (string, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			return imp.TagID, nil
+		}
+	}
+
+	return "", errImpNotFound
+}
+
+func (a *adapter) createBannerAd(bid openrtb2.Bid, ext responseExt, request *openrtb2.BidRequest, seat string) (string, error) {
+	var mcad mcAd
+
+	if strings.Contains(bid.AdM, "<!--preformatted-->") {
+		// Banner ad is already formatted
+		return bid.AdM, nil
+	}
+
+	// create McAd payload
+	mcad.Id = request.ID
+	mcad.Seat = seat
+	mcad.SeatBid = []openrtb2.SeatBid{
+		{Bid: []openrtb2.Bid{bid}},
+	}
+
+	mcMarshalled, err := json.Marshal(mcad)
+	if err != nil {
+		return "", err
+	}
+
+	bannerData := &templatePayload{
+		SiteId:  ext.SiteId,
+		SlotId:  ext.SlotId,
+		AdLabel: ext.AdLabel,
+		PubId:   ext.PublisherId,
+		Page:    request.Site.Page,
+		Referer: request.Site.Ref,
+		McAd:    base64.URLEncoding.EncodeToString(mcMarshalled),
+	}
+
+	var filledTemplate bytes.Buffer
+	if err := a.bannerTemplate.Execute(&filledTemplate, bannerData); err != nil {
+		return "", err
+	}
+
+	return filledTemplate.String(), nil
+}
+
+func getImpSize(imp openrtb2.Imp) string {
+	if imp.Banner == nil || len(imp.Banner.Format) == 0 {
+		return impFallbackSize
+	}
+
+	var (
+		areaMax int64
+		impSize = impFallbackSize
+	)
+
+	for _, size := range imp.Banner.Format {
+		area := size.W * size.H
+		if area > areaMax {
+			areaMax = area
+			impSize = fmt.Sprintf("%dx%d", size.W, size.H)
+		}
+	}
+
+	return impSize
+}
+
+// getBidParameters reads additional data for this imp (site id , placement id, test)
+// Errors in parameters do not break imp flow, and thus are not returned
+func getBidParameters(imp openrtb2.Imp) openrtb_ext.ExtImpSspbc {
+	var extBidder adapters.ExtImpBidder
+	var extSSP openrtb_ext.ExtImpSspbc
+
+	if err := json.Unmarshal(imp.Ext, &extBidder); err == nil {
+		_ = json.Unmarshal(extBidder.Bidder, &extSSP)
+	}
+
+	return extSSP
+}
+
+// getRequestType checks what kind of request we have. It can either be:
+// - a standard request, where all Imps have complete site / placement data
+// - a oneCodeRequest, where site / placement data has to be determined by server
+// - a test request, where server returns fixed example ads
+func getRequestType(request *openrtb2.BidRequest) int {
+	incompleteImps := 0
+
+	for _, imp := range request.Imp {
+		// Read data for this imp
+		extSSP := getBidParameters(imp)
+
+		if extSSP.IsTest != 0 {
+			return requestTypeTest
+		}
+
+		if extSSP.SiteId == "" || extSSP.Id == "" {
+			incompleteImps += 1
+		}
+	}
+
+	if incompleteImps > 0 {
+		return requestTypeOneCode
+	}
+
+	return requestTypeStandard
+}
+
+func formatSspBcRequest(a *adapter, request *openrtb2.BidRequest) (*openrtb2.BidRequest, error) {
```

Comment:
The `a` parameter is not used in this method.

Reply:
Thanks.